### DISCUSSION TITLE
How to implement POST request with different validation and logic depending on media type of request body

### DISCUSSION THREAD

**Discussion by Sh4pe** [fastapi/fastapi#8765] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# No code
```

### Description

I'd like to implement a path operation function for a `POST` request that can take either an `application/json` or `text/plain` request body. In case of the `application/json` response, I'd like to use the "normal" FastAPI validation techniques, but for `text/plain` I'd like to implement my own validation. The path operation function will likely have to use slightly different logic depending on the media type of the request body.

In OpenAPI, you can define e.g. different schemas depending on the media type of the request body, see [Request Body Object](https://swagger.io/specification/#request-body-object) - `content` is a map of media type to "Media Type Object". Of course, I could just add `request: Request` to the parameters of the path operation function and implement the logic myself. However, that would probably not find its way into the OpenAPI docs at `/docs`.

Is there a way to define the path operation function in a way so that I can define different validation and logic for different media types?

### Operating System

Linux, Windows, macOS

### FastAPI Version

0.66.0

### Python Version

3.9.5

**Comment by AlexanderPodorov** [comment, marked as answer]
@Sh4pe , as far as I know it is not possible. I see two options:
1. make two different functions, for `application/json` and `text/plain`. As you mentioned the validation is expected to be pretty different, that makes sense.
2. Try to wrap your original `application/json` and `text/plain` request into one request, like that:
```python
class MyResponse(BaseModel):
    json: Optional[Any] #your original json request
    text: Optional[str] #your original text request
```
We can add a validator to `pydantic` model (see: https://pydantic-docs.helpmanual.io/usage/validators/). So validation should be done pretty easy.
